### PR TITLE
Add Sapling fields to JSON RPC output using TxToJSON.

### PR DESCRIPTION
**Pull request by bitcartel**
Related to #3063 to add Sapling support to wallet RPCs.

### CODE REVIEW

File: src/rpc/rawtransaction.cpp [hunk 1]
```diff
@@ -114,6 +114,38 @@ UniValue TxJoinSplitToJSON(const CTransaction& tx) {
...
+        obj.push_back(Pair("authSig", HexStr(spendDesc.spendAuthSig.begin(), spendDesc.spendAuthSig.end())));
+        vdesc.push_back(obj);
+    }
+    return vdesc;
+}
+
+UniValue TxShieldedOutputsToJSON(const CTransaction& tx) {
+    UniValue vdesc(UniValue::VARR);
+    for (unsigned int i = 0; i < tx.vShieldedOutput.size(); i++) {
+        const OutputDescription& outputDesc = tx.vShieldedOutput[i];
```
Comment (str4d) [inline src/rpc/rawtransaction.cpp]: Nit: `for (const OutputDescription& outputDesc : tx.vShieldedOutput) {`

File: src/rpc/rawtransaction.cpp [hunk 2]
```diff
@@ -160,6 +192,17 @@ void TxToJSON(const CTransaction& tx, const uint256 hashBlock, UniValue& entry)
     UniValue vjoinsplit = TxJoinSplitToJSON(tx);
     entry.push_back(Pair("vjoinsplit", vjoinsplit));
 
+    if (tx.fOverwintered && tx.nVersion >= SAPLING_TX_VERSION) {
+        entry.push_back(Pair("valuebalance", ValueFromAmount(tx.valueBalance)));
```
Comment (str4d) [inline src/rpc/rawtransaction.cpp]: `valueBalance`

File: src/rpc/rawtransaction.cpp [hunk 3]
```diff
@@ -160,6 +192,17 @@ void TxToJSON(const CTransaction& tx, const uint256 hashBlock, UniValue& entry)
...
+        UniValue vspenddesc = TxShieldedSpendsToJSON(tx);
+        entry.push_back(Pair("vshieldedspends", vspenddesc));
```
Comment (str4d) [inline src/rpc/rawtransaction.cpp]: Either `vShieldedSpend` or `shieldedSpends`.

File: src/rpc/rawtransaction.cpp [hunk 4]
```diff
@@ -160,6 +192,17 @@ void TxToJSON(const CTransaction& tx, const uint256 hashBlock, UniValue& entry)
...
+        UniValue voutputdesc = TxShieldedOutputsToJSON(tx);
+        entry.push_back(Pair("vshieldedoutputs", voutputdesc));
```
Comment (str4d) [inline src/rpc/rawtransaction.cpp]: Either `vShieldedOutput` or `shieldedOutputs`.

File: src/rpc/rawtransaction.cpp [hunk 5]
```diff
@@ -160,6 +192,17 @@ void TxToJSON(const CTransaction& tx, const uint256 hashBlock, UniValue& entry)
...
+        if (!(vspenddesc.size()==0 && voutputdesc.size()==0)) {
```
Comment (str4d) [inline src/rpc/rawtransaction.cpp]: `if (!(vspenddesc.empty() && voutputdesc.empty())) {`

File: src/rpc/rawtransaction.cpp [hunk 6]
```diff
@@ -160,6 +192,17 @@ void TxToJSON(const CTransaction& tx, const uint256 hashBlock, UniValue& entry)
     UniValue vjoinsplit = TxJoinSplitToJSON(tx);
     entry.push_back(Pair("vjoinsplit", vjoinsplit));
 
+    if (tx.fOverwintered && tx.nVersion >= SAPLING_TX_VERSION) {
```
[the remark this answers is not in the document]
Comment (bitcartel) [reply src/rpc/rawtransaction.cpp]: I think it's fine too (can always change later).  Currently txs print out empty vjoinsplit arrays.

File: src/rpc/rawtransaction.cpp [hunk 7]
```diff
@@ -114,6 +114,38 @@ UniValue TxJoinSplitToJSON(const CTransaction& tx) {
     return vjoinsplit;
 }
 
+UniValue TxShieldedSpendsToJSON(const CTransaction& tx) {
+    UniValue vdesc(UniValue::VARR);
+    for (unsigned int i = 0; i < tx.vShieldedSpend.size(); i++) {
+        const SpendDescription& spendDesc = tx.vShieldedSpend[i];
```
[the remark this answers is not in the document]
Comment (bitcartel) [reply src/rpc/rawtransaction.cpp]: Yeah, we don't need `i` for anything in the output

File: src/rpc/rawtransaction.cpp (hunk 1 above)
Comment (bitcartel) [reply src/rpc/rawtransaction.cpp]: ok

**Comment by bitcartel** [comment]
@Eirik0 @str4d Pushed changes for nits and updated qa test.

**Comment by Eirik0** [comment]
utACK

**Comment by str4d** [comment]
@zkbot r+

**Pull request merged**